---
title: "Data import and manipulation in poppr version `r packageVersion('poppr')`"
author: Zhian N. Kamvar and Niklaus J. Grünwald
date: "`r Sys.Date()`"
bibliography: the_bibliography.bib
output: 
  rmarkdown::html_vignette:
    fig_caption: yes
    fig_width: 5
    fig_height: 5
    toc: true
    toc_depth: 2
vignette: >
  %\VignetteIndexEntry{Data Import and Manipulation}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, echo = FALSE, message = FALSE, warning = FALSE}
knitr::opts_chunk$set(message = FALSE, warning = FALSE, tidy = FALSE)
knitr::opts_chunk$set(fig.align = "center", fig.show = 'asis')
library("knitcitations")
cite_options(citation_format = "pandoc", max.names = 3, style = "html", 
             hyperlink = "to.doc")
bib <- read.bibtex("the_bibliography.bib")
```

```{r poppr_funk, eval = TRUE, echo = FALSE}
print_command <- function(funk){
  fargs <- formals(funk)

  lapply(names(fargs), function(arg_name, fargs){
    arg <- fargs[[arg_name]]
    if (missing(arg)){
      fargs[[arg_name]] <<- as.symbol(arg_name)
      names(fargs)[names(fargs) == arg_name] <<- ""
    }
  }, fargs)
  fargs$call <- as.symbol(funk)
  fargs <- fargs[c(length(fargs), 1:(length(fargs) - 1))]
  return(as.call(fargs))
}
```


<center>

Abstract
========

*Poppr* provides open-source, cross-platform tools for quick analysis of
population genetic data enabling focus on data analysis and interpretation.
While there are a plethora of packages for population genetic analysis, few are
able to offer quick and easy analysis of populations with mixed reproductive
modes. *Poppr*'s main advantage is the ease of use and integration with other
packages such as *adegenet* and *vegan*, including support for novel methods
such as clone correction, multilocus genotype analysis, calculation of Bruvo's
distance, and the index of association. New features in version 2.0 include
generation of minimum spanning networks with reticulation, calculation of the
index of association for genomic data, and filtering multilocus genotypes based
on genetic distance.

<br>
<br>

<img src="jalapeno_logo.png" width="300">
</center>

Introduction {#intro}
============

Purpose {#intro:purpose}
-------

*Poppr* is an R package with convenient functions for analysis of genetic data
with mixed modes of reproduction including sexual and clonal reproduction. While
there are many R packages in CRAN and other repositories with tools for
population genetic analyses, few are appropriate for populations with mixed
modes of reproduction. There are several stand-alone programs that can handle
these types of data sets, but they are often platform specific and often only
accept specific data types. Furthermore, a typical analysis often involves
switching between many programs, and converting data to each specific format.

*Poppr* is designed to make analysis of populations with mixed reproductive
modes more streamlined and user friendly so that the researcher using it can
focus on data analysis and interpretation. *Poppr* allows analysis of haploid
and diploid dominant/co-dominant marker data including microsatellites, Single
Nucleotide Polymorphisms (SNP), and Amplified Fragment Length Polymorphisms
(AFLP). To avoid creating yet another file format that is specific to a program,
*poppr* was created on the backbone of the popular R package *adegenet* and can
take all the file formats that *adegenet* can take (Genpop, Genetix, Fstat, and
Structure) and newly introduces compatibility with GenAlEx formatted files
(exported to CSV). This means that anything you can analyze in *adegenet* can be
further analyzed with *poppr*.

The real power of *poppr* is in the data manipulation and analytic tools.
*Poppr* has the ability to bootstrap, clone-censor, and subset data sets. With
*poppr* you can also quickly calculate Bruvo's distance, the index of
association, and easily determine which multilocus genotypes are shared across
populations.

In version 2.0, tools for genomic data were introduced [@kamvar2015novel]. These
tools allow researchers to define what it means to be a clone lost in a sea of
genomic data, generate bootstrapped dendrograms with any genetic distance, and
calculate minimum spanning networks with reticulations to reveal the underlying
population structure of your clonal data.

Resources
---------

This vignette will cover all of the material you need to know to efficiently
analyze data in *poppr*. For information on methods of analysis (eg. index of
association, distance measures, AMOVA, ...), please read the manual pages 
provided for each function.

As *poppr* expanded from version 1.0, the vignette also expanded to be 80+
pages. As a result, it became clear that over 22,000 words was less of a manual
and more of a novella with a terrible plot. To remedy this, this vignette will
focus only on data manipulation and a separate vignette, "algo", has been
written to give algorithmic details of analyses introduced with *poppr*.

As of spring 2014, Drs. Niklaus J. Grünwald, Sydney E. Everhart, and I have co-
authored a primer on using R for population genetic analysis. It is located at
<http://grunwaldlab.github.io//Population_Genetics_in_R> and the source code can
be found [on our github site][githubsite].

Getting Help
------------

If you have any questions or feedback, feel free to send a message to
the *poppr* forum at <http://groups.google.com/group/poppr>. You can submit bug
reports there or on our github site: 
<https://github.com/grunwaldlab/poppr/issues>

Acknowledgments
---------------

Much thanks goes to Sydney E. Everhart for alpha testing, beta testing,
feature requests, proofreading, data contribution, and moral support
throughout the writing of this package and manual. Thanks also to Brian
Knaus, Ignazio Carbone, David Cooke, Corine Schoebel, Jane Stewart, and
Zaid Abdo for beta testing and feedback.

The following data sets are included in *poppr*:

-   `Pinf` (SSR, *Phytophthora infestans*)[@goss2014irish]

-   `monpop` (SSR, *Monilinia fructicola*)[@everhart2014finescale]

-   `Aeut` (AFLP, *Aphanomyces eutieches*)[@Grunwald:2006]

-   `Pram` (SSR, *Phytophthora
    ramorum*)[@kamvar2015spatial; @goss2009population]

Citation
--------

To cite *poppr*, please type in your R console:

```{r popprcite, eval = TRUE}
citation(package = "poppr")
```

The formal publication for the first version of *poppr* was
published in the journal **PeerJ**: <http://peerj.com/articles/281/>
[@kamvar2014poppr]. The new features in version 2.0 were published in
**Frontiers** <http://dx.doi.org/10.3389/fgene.2015.00208>
[@kamvar2015novel].

Installation and Quick Start
============================

Installation {#intro:install}
------------

This manual assumes you have installed R. If you have not, please refer
to The CRAN home page at <http://cran.r-project.org/>. We also recommend
the Rstudio IDE (<http://www.rstudio.com/>), which allows the user to
view the R console, environment, scripts, and plots in a single window.

### From CRAN

To install *poppr* from CRAN, select "Package Installer"
from the menu "Packages & Data" in the GUI or type:

```{r install, eval = FALSE}
install.packages("poppr", repos = "http://cloud.r-project.org/")
```

All dependencies will also be installed. In the unfortunate case this does not 
work, please consult
<http://cran.r-project.org/doc/manuals/R-admin.html#Installing-packages>.

### From GitHub

GitHub is a repository where you can find all stable and development versions of
*poppr*.

Since *poppr* contains C code, it needs to be compiled, which means that you
need a working C compiler. If you are on Linux, you should have that, but if you
are on Windows or OSX, you might need to download some special tools:

Windows

:   Download Rtools: <http://cran.r-project.org/bin/windows/Rtools/>

OSX

:   Download Xcode: <https://developer.apple.com/xcode>

To install from GitHub, you do not need to download the tarball since
there is a package called *devtools* that will download and install the
package for you directly from GitHub. After you have installed all
dependencies (see above section), you should download *devtools*:

```{r install_devtools, eval = FALSE}
install.packages("devtools")
```

Now you can execute the command `install_github()` with the user and
repository name:

```{r install_github, eval = FALSE}
devtools::install_github("grunwaldlab/poppr")
```

Quick start {#intro:qstart}
-----------

The author assumes that if you have reached this point in the manual, then you
have successfully installed R and *poppr*. Before proceeding, you should be
aware that R is case sensitive. This means that the words "Case" and "case" are
different. You should also know where your R package library is located. In this
section, you will learn how to locate a file, import it to R, and make a first
analysis using the `poppr()` function.

> **What or where is my R package library?** 
>
> R is as powerful as it is through a community of people who submit extra code
> called ``packages" to help it do specific things. These packages live in a
> certain place on your computer called an R library. You can find out where
> this library is by typing `.libPaths()`

Importing a file into R involves you knowing the path to your file and then
typing that into R's console. `getfile()` will help provide a point and click
interface for selecting a file. This is simply a tool to help you get started.
As you get better with R, you might feel that you don"t need it at all.

First, tell your computer to search R's library, load the *poppr* package, and
use `getfile()`:

```{r echo = FALSE}
x <- list(files = "/path/to/R/poppr/files/rootrot.csv", path = "/path/to/R/poppr/files")
```

```{r message = TRUE}
library("poppr")
```

```{r getfilefunk, eval=FALSE}
x <- getfile()
```

A pop up window will appear like this[^1]:

![A popup window as it appears in OSX (Mountain Lion).<span
data-label="getfile window"></span>](getfile.png)

You should navigate to your R library and select the file called
"rootrot.csv". If you don"t know where your the poppr package lives, you
can find it by typing `find.package("poppr")` into the R command line.
Once we select a file, the file name and its path will be stored in the
variable, `x`. We can confirm that by typing `x` into R"s command line.

```{r getfilex}
x
```

Here we can see that `x` is a list with two entries: `$files` shows the
files you selected and `$path` shows the path to those files.

We will use `x$files` to access the file. The file is in the GenAlEx format, so
we will import it using `read.genalex()` and then analyze it with the function
`poppr()` to get a table of diversity indices per population.


```{r firstpoppr, eval=FALSE}
myData  <- read.genalex(x$files)
myData
```

```{r echo = FALSE}
options(width=90)
myData <- read.genalex(system.file("files/rootrot.csv", package="poppr"))
```

```{r aflp}
popdata <- poppr(myData)
```

The output of `poppr()` was assigned to the variable `popdata`, so let"s
look at the data.

```{r firstpoppr2}
popdata
```


The fields you see in the output include:

-   `Pop -` Population name (Note that "Total" also means "Pooled").

-   `N -` Number of individuals observed.

-   `MLG -` Number of multilocus genotypes (MLG) observed.

-   `eMLG -` The number of expected MLG at the smallest sample size
    $\geq 10$ based on rarefaction. [@Hurlbert:1971]

-   `SE -` Standard error based on `eMLG` [@Heck:1975]

-   `H -` Shannon-Wiener Index of MLG diversity. [@Shannon:1948]

-   `G -` Stoddart and Taylor"s Index of MLG diversity. [@Stoddart:1988]

-   `lambda -` Simpson"s index, $\lambda$.

-   `E.5 -` Evenness, $E_5$.
    [@Pielou:1975][@Ludwig:1988][@Grunwald:2003]

-   `Hexp -` Nei"s 1978 Expected Heterozygosity. [@Nei:1978]

-   `Ia -` The index of association, $I_A$. [@Brown:1980] [@Smith:1993]
    [@Agapow:2001]

-   `rbarD -` The standardized index of association, $\bar{r}_d$.
    [@Agapow:2001]

These fields are further described in the function `poppr`. You can
access the help page for `poppr` by typing `?poppr` in your R console.

One thing to note about this output is the `NaN` in the column labeled
`SE`. In R, `NaN` means "Not a number". This is produced from
calculation of a standard error based on rarefaction analysis.
Occasionally, this calculation will encounter a situation in which it
must attempt to take a square root of a negative number. Since the root
of any negative number is not defined in the set of real numbers, it
must therefore have an imaginary component, $i$. Unfortunately, R will
not represent the imaginary components of numbers unless you
specifically tell it to do so. By default, R represents these as `NaN`.

[githubsite]: https://github.com/grunwaldlab/Population_Genetics_in_R